File: kth_smallest_element_bst.py
# Brute Force Approach: Perform inorder traversal and return element at k- 1 index as value
# Time - O(N) traversal on all roots
# Space - O(N)


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
    
        inOrderTrav = self.inOrder(root)
        #print(inOrderTrav)
        return inOrderTrav[k-1]
        
        
    def inOrder(self, root):
        
        if root == None:
            return []
        
        return self.inOrder(root.left) + [root.val] + self.inOrder(root.right)
        

## Approach 2: Iterative DFS and stop when k becomes zero
# Time - O(N)
# Space - O(H)      


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        
        if root == None:
            return 0
        
        stack = []
        
        while stack != [] or root != None:
            
            while root != None:
                stack.append(root)
                root = root.left
            
            root = stack.pop()
            k -= 1
            
            if not k:
                return root.val
            
            root = root.right
                
            
         
# Approach 3: DFS Recursive solution with optimization to suppress recursion(ie stop once we find the kth value)
# Time - O(N)
# Space - O(H) height of Tree is the number of nodes in the stack

def kthSmallest(self, root: TreeNode, k: int) -> int:
        
    self.k = k
    self.result = None
    self.kthSmallestHelper(root, k)
    
    return self.result
     
def kthSmallestHelper(self, root, k):
    # base case
    if root == None:
        return
    
    #logic
    self.kthSmallestHelper(root.left, k)
    
    self.k -= 1
    
    if self.k == 0:
        self.result = root.val
        return
    
    if self.result == None: # if we still havent found the result only then proceed to right side of tree
        self.kthSmallestHelper(root.right, k)       

File: problem1/KthSmallestElementBST.java
// Time Complexity : O(n), n -> Number of nodes in the tree
// Space Complexity : O(log n), log n -> Height of Tree
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
package problem1;

public class KthSmallestElementBST {
	int result;
	int cnt;

	/********************* Inorder Traversal - Recursive *********************/
	public int kthSmallest(TreeNode root, int k) {
		if (root == null) {
			return -1;
		}

		cnt = 0;

		getInorder(root, k);
		return result;
	}

	private void getInorder(TreeNode root, int k) {
		if (root == null) {
			return;
		}

		getInorder(root.left, k);
		cnt++;
		if (cnt == k) {
			result = root.val;
			return;
		}
		getInorder(root.right, k);
	}

	public static void main(String[] args) {
		KthSmallestElementBST obj = new KthSmallestElementBST();

		TreeNode root = new TreeNode(6);
		root.left = new TreeNode(2);
		root.right = new TreeNode(8);

		root.left.left = new TreeNode(0);
		root.left.right = new TreeNode(4);

		root.left.right.left = new TreeNode(3);
		root.left.right.right = new TreeNode(5);

		root.right.left = new TreeNode(7);
		root.right.right = new TreeNode(9);

		int k = 4;

		System.out.println("Smallest element number \'" + k + "\' = " + obj.kthSmallest(root, k));
	}

}
